File: tools/pdf-batch-editor/process-partner-logos.py
import json
from pathlib import Path
from PIL import Image

ROOT = Path(r"C:\Users\Administrator\Desktop\cnwslchain-website")
RAW = ROOT / "tools/pdf-batch-editor/_logos/raw"
OUT = ROOT / "public/images/home/partners"
LOGS = Path(r"C:\Users\Administrator\.cursor\browser-logs")

WIDTH, HEIGHT, PAD = 640, 280, 48
OUT.mkdir(parents=True, exist_ok=True)


def extract_cdp(json_name: str, out_name: str) -> None:
    payload = json.loads((LOGS / json_name).read_text(encoding="utf-8"))
    b64 = payload.get("data") or payload.get("result", {}).get("data")
    if not b64:
        raise SystemExit(f"No image data in {json_name}")
    import base64

    (RAW / out_name).write_bytes(base64.b64decode(b64))


extract_cdp("cdp-response-Page.captureScreenshot-2026-09-05T08-31-23-360Z.json", "genesis-clip.png")
extract_cdp("cdp-response-Page.captureScreenshot-2026-09-05T08-33-06-822Z.json", "foxconn-clip.png")
extract_cdp("cdp-response-Page.captureScreenshot-2026-09-05T08-31-16-748Z.json", "avic-clip.png")

jobs = [
    ("bambu-lab", "bambu-favicon.ico", False),
    ("huawei", "huawei-official.png", False),
    ("hapm-magna", "hapm-magna-header.png", False),
    ("trumpchi", "trumpchi-logo.png", True),
    ("foxconn", "foxconn-clip.png", True),
    ("genesis", "genesis-clip.png", True),
    ("gac", "gac-header.png", False),
    ("avic-jingji", "avic-clip.png", False),
    ("midea", "midea-official.png", False),
]


def knockout_light_logo(image: Image.Image) -> Image.Image:
    """Turn white wordmarks on dark/black clips into dark marks on transparency."""
    pixels = image.load()
    for y in range(image.height):
        for x in range(image.width):
            r, g, b, a = pixels[x, y]
            lum = 0.299 * r + 0.587 * g + 0.114 * b
            if a < 24 or lum < 165:
                pixels[x, y] = (0, 0, 0, 0)
            else:
                strength = min(255, int(a * (lum - 165) / 90))
                pixels[x, y] = (28, 36, 48, max(0, strength))
    return image


def fit(src: Image.Image) -> tuple[int, int, int, int]:
    box_w, box_h = WIDTH - PAD * 2, HEIGHT - PAD * 2
    scale = min(box_w / src.width, box_h / src.height)
    w = max(1, round(src.width * scale))
    h = max(1, round(src.height * scale))
    x = (WIDTH - w) // 2
    y = (HEIGHT - h) // 2
    return x, y, w, h


for logo_id, src_name, invert in jobs:
    src_path = RAW / src_name
    image = Image.open(src_path)
    image = image.convert("RGBA")
    if invert:
        image = knockout_light_logo(image)
    canvas = Image.new("RGB", (WIDTH, HEIGHT), (255, 255, 255))
    x, y, w, h = fit(image)
    resized = image.resize((w, h), Image.Resampling.LANCZOS)
    canvas.paste(resized, (x, y), resized)
    dest = OUT / f"{logo_id}.webp"
    canvas.save(dest, "WEBP", quality=90, method=6)
    print(f"{logo_id}\t{image.size}\t-> {w}x{h}\t{dest.stat().st_size}")

File: tools/pdf-batch-editor/process-partner-logos.mjs
import fs from "fs";
import path from "path";
import { createCanvas, loadImage } from "@napi-rs/canvas";

const ROOT = "C:/Users/Administrator/Desktop/cnwslchain-website";
const RAW = path.join(ROOT, "tools/pdf-batch-editor/_logos/raw");
const OUT = path.join(ROOT, "public/images/home/partners");
const LOGS = "C:/Users/Administrator/.cursor/browser-logs";

const WIDTH = 640;
const HEIGHT = 280;
const PAD = 48;

function extractCdpPng(jsonName, outName) {
  const json = JSON.parse(fs.readFileSync(path.join(LOGS, jsonName), "utf8"));
  const b64 = json.data || json.result?.data;
  if (!b64) throw new Error(`No image data in ${jsonName}`);
  const dest = path.join(RAW, outName);
  fs.writeFileSync(dest, Buffer.from(b64, "base64"));
  return dest;
}

extractCdpPng("cdp-response-Page.captureScreenshot-2026-09-05T08-31-23-360Z.json", "genesis-clip.png");
extractCdpPng("cdp-response-Page.captureScreenshot-2026-09-05T08-33-06-822Z.json", "foxconn-clip.png");
extractCdpPng("cdp-response-Page.captureScreenshot-2026-09-05T08-31-16-748Z.json", "avic-clip.png");

const jobs = [
  { id: "bambu-lab", src: "bambu-official.svg" },
  { id: "huawei", src: "huawei-official.png" },
  { id: "hapm-magna", src: "hapm-magna-header.png" },
  { id: "trumpchi", src: "trumpchi-logo.png" },
  { id: "foxconn", src: "foxconn-clip.png" },
  { id: "genesis", src: "genesis-clip.png" },
  { id: "gac", src: "gac-header.png" },
  { id: "avic-jingji", src: "avic-clip.png" },
  { id: "midea", src: "midea-official.png" },
];

fs.mkdirSync(OUT, { recursive: true });

function fitRect(srcW, srcH, boxW, boxH) {
  const scale = Math.min(boxW / srcW, boxH / srcH);
  const w = Math.round(srcW * scale);
  const h = Math.round(srcH * scale);
  return {
    w,
    h,
    x: Math.round((WIDTH - w) / 2),
    y: Math.round((HEIGHT - h) / 2),
  };
}

for (const job of jobs) {
  const srcPath = path.join(RAW, job.src);
  const image = await loadImage(srcPath);
  const canvas = createCanvas(WIDTH, HEIGHT);
  const ctx = canvas.getContext("2d");
  ctx.fillStyle = "#ffffff";
  ctx.fillRect(0, 0, WIDTH, HEIGHT);
  const box = fitRect(image.width, image.height, WIDTH - PAD * 2, HEIGHT - PAD * 2);
  ctx.drawImage(image, box.x, box.y, box.w, box.h);
  const outPath = path.join(OUT, `${job.id}.webp`);
  fs.writeFileSync(outPath, await canvas.encode("webp", 90));
  console.log(`${job.id}\t${image.width}x${image.height}\t-> ${box.w}x${box.h}\t${fs.statSync(outPath).size}`);
}
